compiler/executable/match_/planner/plan.rs

```diff
@@ -941,42 +949,47 @@ impl PartialCostPlan {
                     graph.elements[&pattern_id].is_valid(pattern_id, &all_available_vars, graph)
                 }
             })
-            .flat_map(move |&extension| {
-                let join_var = self.determine_joinability(graph, extension);
-
-                if join_var.is_none() {
-                    vec![(extension, join_var)].into_iter()
-                } else {
-                    vec![(extension, None), (extension, join_var)].into_iter()
-                }
+            .flat_map(move |&extension| match self.determine_joinability(graph, extension) {
+                None => vec![(extension, None)].into_iter(),
+                Some(var) => vec![(extension, None), (extension, Some(var))].into_iter(),
             })
-            .map(move |(extension, join_var)| {
-                let added_cost: Cost;
-                let meta_data: CostMetaData;
-
-                if join_var.is_none() {
-                    (added_cost, meta_data) =
-                        self.compute_added_cost(graph, extension, &all_available_vars, join_var)?;
+            .map(move |(extension, join_var_opt)| {
+                let planner_vertex = &graph.elements[&VertexId::Pattern(extension)];
+                let (cost_before_extension, cost_of_extension, metadata) = if let Some(join_var) = join_var_opt {
+                    // Step continues
+                    debug_assert!(matches!(planner_vertex, PlannerVertex::Constraint(_)));
```

Comment:
This assert doubles the check below with no additional information. Either remove or add a message with the failing constraint.